### PR TITLE
feat!: add output variable estimation for `ScriptCallHandler`

### LINKED ISSUE #1001: Add tx dependency estimation to script calls
Script calls should have tx dependency estimation just like contract calls

### PR DESCRIPTION
**Pull request by MujkicA**
Closes #1001 

Adds tx dep estimation to script calls. I also attempted to deduplicate some code by introducing a new trait `TxDependencyExtension`. The only caveat I see with it is that the `ViewOnlyAccount` trait has to enforce a `'static` trait bound i.e. not contain non-static references. Let me know what you think.

Thanks to @hal3e for helping make the trait neat and clean.

BREAKING CHANGE:  `TxDependencyExtension` needs to be in scope to use `append_variable_outputs` and `append_contract`

### Checklist
- [x] I have linked to any relevant issues.
- [x] I have updated the documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary labels.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Review by hal3e — COMMENTED** [review]
Looks very good. Did you forget to add an `estimation` test ?

**Review by Br1ght0ne — APPROVED** [review]
LGTM, solid deduplication effort too. Re `ViewOnlyAccount` getting a `'static` bound: I had this PR compiling and tested with `'static` removed - what was the reasoning behind it?
Leaving an approval as this note is not significant.

File: packages/fuels-programs/src/call_utils.rs [hunk 1]
```diff
@@ -30,6 +31,35 @@ pub(crate) struct CallOpcodeParamsOffset {
     pub call_data_offset: usize,
 }
 
+/// How many times to attempt to resolve missing tx dependencies.
+pub const DEFAULT_TX_DEP_ESTIMATION_ATTEMPTS: u64 = 10;
+
+#[async_trait::async_trait]
+pub trait TxDependencyEstimation: Sized {
```
Comment (iqdecay) [inline packages/fuels-programs/src/call_utils.rs:38]: nice one with the default impl

File: packages/fuels-programs/src/script_calls.rs [hunk 2]
```diff
@@ -184,7 +248,11 @@ where
     async fn call_or_simulate(&mut self, simulate: bool) -> Result<FuelCallResponse<D>> {
         let chain_info = self.provider.chain_info().await?;
         let tb = self.prepare_builder().await?;
-        let tx = self.account.add_fee_resources(tb, 0, None).await?;
+        let base_amount = self.calculate_base_asset_sum();
+        let tx = self
+            .account
+            .add_fee_resources(tb, base_amount, None)
```
Comment (iqdecay) [inline packages/fuels-programs/src/script_calls.rs:254]: nice catch

File: packages/fuels-programs/src/script_calls.rs [hunk 3]
```diff
@@ -246,3 +314,32 @@ where
...
+            let contract_id = Bech32ContractId::from(
+                *panic_receipt
+                    .contract_id()
+                    .expect("Panic receipt must contain contract id."),
+            );
+            self = self.append_contract(contract_id);
+        }
+
+        self
+    }
```
Comment (iqdecay) [inline packages/fuels-programs/src/script_calls.rs:344]: why cannot this method have its default implementation inside the trait? It looks like it's the same for both `ScriptCall` and `ContractCall`, and can be overidden for `MultiCall`.

**Review by iqdecay — APPROVED** [review]
left one question but it's non-blocking.

**Comment by hal3e** [comment]
> LGTM, solid deduplication effort too.
> 
> Re `ViewOnlyAccount` getting a `'static` bound: I had this PR compiling and tested with `'static` removed - what was the reasoning behind it?
> 
> Leaving an approval as this note is not significant.

Excellent catch! The tx estimation trait was changed a bit so now the `'static` is not necessary anymore.

File: packages/fuels-programs/src/call_utils.rs [hunk 4]
```diff
@@ -405,6 +435,18 @@ fn extract_unique_contract_ids(calls: &[ContractCall]) -> HashSet<ContractId> {
...
+    receipts.iter().any(
+        |r| matches!(r, Receipt::Revert { ra, .. } if *ra == FAILED_TRANSFER_TO_ADDRESS_SIGNAL),
+    )
+}
+
+pub fn find_contract_not_in_inputs(receipts: &[Receipt]) -> Option<&Receipt> {
+    receipts.iter().find(
+        |r| matches!(r, Receipt::Panic { reason, .. } if *reason.reason() == PanicReason::ContractNotInInputs ),
+    )
+}
```
Comment (segfault-magnet) [inline packages/fuels-programs/src/call_utils.rs:448]: It seems all callers of this function proceed to do the same thing -- extract and map the contract_id. 

We can make the helper a bit more helpful: 
```rust
pub fn find_id_of_missing_contract(receipts: &[Receipt]) -> Option<Bech32ContractId> {
    receipts.iter().find_map(|receipt| match receipt {
        Receipt::Panic {
            reason,
            contract_id,
            ..
        } if *reason.reason() == PanicReason::ContractNotInInputs => {
            let contract_id = contract_id
                .expect("panic caused by a contract not in inputs must have a contract id");
            Some(Bech32ContractId::from(contract_id))
        }
        _ => None,
    })
}
```

This also avoids the double check whether we're dealing with a Panic receipt (check hidden in the `contract_id` getter).

**Review by segfault-magnet — DISMISSED** [review]

**Review by digorithm — APPROVED** [review]
Super neat trait you've got here 🤌

:shipit:  -- though I agree with the other reviewers here w.r.t. some non-urgent non-blocking improvements.

**Comment by hal3e** [comment]
@FuelLabs/sdk-rust thansk for the comments! I implemented all the suggestions you left. 
@iqdecay's idea to make a default `append_missing_deps` led to some really nice de-duplication.

File: packages/fuels-programs/src/call_utils.rs [hunk 5]
```diff
@@ -35,9 +35,45 @@ pub(crate) struct CallOpcodeParamsOffset {
...
+    /// pairs and set them into the transaction. Note that this is a builder
+    /// method, i.e. use it as a chain:
+    ///
+    /// ```ignore
+    /// my_contract_instance.my_method(...).append_contracts(additional_contract_id).call()
+    /// my_script_instance.main(...).append_contracts(additional_contract_id).call()
+    /// ```
+    ///
+    /// [`Input::Contract`]: fuel_tx::Input::Contract
+    /// [`Output::Contract`]: fuel_tx::Output::Contract
```
Comment (iqdecay) [inline packages/fuels-programs/src/call_utils.rs:64]: Should use the name `append_contract` not `append_contracts`

**Review by iqdecay — APPROVED** [review]
one nit in docs

**Review by segfault-magnet — APPROVED** [review]
Took the liberty of merging the latest master in and fixing a typo so that I may approve and unblock you.

**Review by Br1ght0ne — APPROVED** [review]

**Review by iqdecay — APPROVED** [review]
lgtm!

**Pull request merged**